cpp/src/replace/replace.cu

```diff
@@ -861,6 +864,137 @@ std::unique_ptr<cudf::column> replace_nulls(cudf::column_view const& input,
 }  // namespace experimental
 }  // namespace cudf
 
+namespace cudf {
+namespace experimental {
+namespace detail {
+namespace {
+
+struct replace_nans_functor {
+  template <typename T, typename Replacement>
```

Comment:
Discovered from `copy_if_else` implementation. I found something more: with the new `make_pair_iterator` for scalar, @karthikeyann did an awesome thing where he made the iterator own the device object. So now you don't have to manage the lifetime of the device view either. I don't quite understand how it works but if it does, then perhaps the same can be used for `column_view` and we wouldn't have to create a `column_device_view` first.

Reply:
Taking another look at this, could this take the `replacement_pair_iterator` as a parameter instead? It might be a little more generic to take an iterator as input rather than types that happen to have the same interfaces.

It seems the `replacement_nullable` parameter and the `if (replacement_nullable) {` branches could be removed (well, refactored out to the call site anyway). The `CUDF_EXPECTS(input.type() == replacement.type(),` would have to be refactored out to the call site as well.

Reply:
Taking another look at this, could this take the replacement_pair_iterator as a parameter instead? It might be a little more generic to take an iterator as input rather than types that happen to have the same interfaces.

It seems the replacement_nullable parameter and the if (replacement_nullable) { branches could be removed (well, refactored out to the call site anyway). The CUDF_EXPECTS(input.type() == replacement.type(), would have to be refactored out to the call site as well.

Reply:
Comparing this to the copy_if functionality, this is a parallel of the `copy_if_else_functor_impl` rather than the iterator based `copy_if_else` API. If I were to make this change, it'd have to be split into a `replace_nans_functor` and an iterator based `replace_nans` function. That iterator based `replace_nans` function would be not much more than a convenience wrapper around `copy_if_else`.

I'd imagine if someone wants to replace nulls in a detail API, they're probably looking for doing that as an intermediate step and not as the final step. Meaning they'd prefer avoiding materializing an intermediate column. That means, it makes more sense to have a `nan_replaced_iterator` (similar to `null_replaced_iterator`). Out of scope for this PR but definitely something that can be done if we find a need for it.